src/libraries/System.Private.CoreLib/src/System/Reflection/MethodInvoker.cs

```diff
@@ -2,78 +2,367 @@
+            scoped Span<bool> shouldCopyBack = stackalloc bool[_argCount];
+
+            regArgStorage = new((void**)pStorage, (uint)_argCount, areByRefs: false);
+            GCFrameRegistration regByRefStorage = new((void**)pByRefStorage, (uint)_argCount, areByRefs: true);
+
+            try
+            {
+                RuntimeImports.RegisterForGCReporting(&regArgStorage);
+                RuntimeImports.RegisterForGCReporting(&regByRefStorage);
+
+                for (int i = 0; i < _argCount; i++)
                 {
-                    throw new TargetInvocationException(e);
+                    object? arg = arguments[i];
+                    shouldCopyBack[i] = CheckArgument(ref arg, i);
+                    copyOfArgs[i] = arg;
+#pragma warning disable CS8500
+                    *(ByReference*)(pByRefStorage + i) = (_invokerArgFlags[i] & InvokerArgFlags.IsValueType) != 0 ?
+#pragma warning restore CS8500
+                        ByReference.Create(ref Unsafe.AsRef<object>(pStorage + i).GetRawData()) :
+                        ByReference.Create(ref Unsafe.AsRef<object>(pStorage + i));
                 }
+
+                object? ret = _invokeFunc_RefArgs!(obj, pByRefStorage);
+                CopyBack(arguments, copyOfArgs, shouldCopyBack);
+                return ret;
             }
-            else if (_invokeFunc != null)
+            finally
             {
-                ret = _invokeFunc(obj, args);
+                RuntimeImports.UnregisterForGCReporting(&regByRefStorage);
+                RuntimeImports.UnregisterForGCReporting(&regArgStorage);
             }
-            else
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        // Copy modified values out. This is only done with ByRef parameters.
+        internal void CopyBack(Span<object?> dest, Span<object?> copyOfParameters, Span<bool> shouldCopyBack)
+        {
+            for (int i = 0; i < dest.Length; i++)
             {
-                ret = InterpretedInvoke(obj, args);
+                if (shouldCopyBack[i])
+                {
+                    if ((_invokerArgFlags[i] & InvokerArgFlags.IsNullableOfT) != 0)
+                    {
+                        Debug.Assert(copyOfParameters[i] != null);
+                        Debug.Assert(((RuntimeType)copyOfParameters[i]!.GetType()).IsNullableOfT);
+                        dest![i] = RuntimeMethodHandle.ReboxFromNullable(copyOfParameters[i]);
+                    }
+                    else
+                    {
+                        dest![i] = copyOfParameters[i];
+                    }
+                }
             }
+        }
 
-            return ret;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool CheckArgument(ref object? arg, int i)
+        {
+            RuntimeType sigType = _argTypes[i];
+
+            // Convert the type if necessary.
+            // Note that Type.Missing is not supported.
```

Comment:
NIT: By assuming this applyes to all new Invoke methods, this might be good to be added to the public doc